### DISCUSSION TITLE
Tracking `MouseFlags.Button1Pressed`...

### DISCUSSION THREAD

**Discussion by tig** [tui-cs/Terminal.Gui#3290] [Q&A]
If the mouse button is pressed. and the View under the mouse doesn't grab the mouse, and then the user moves the mouse over another view, that view gets a mouse event with `MouseFlags.Button1Pressed`. As far as i can tell, with the current API there's no way to tell that the mouse button was presssed *before* the mouse moved over the new view.

Am I missing something?

**Comment by BDisp** [comment]
If the mouse is moving with the Button1Pressed then the ReportMousePosition must exxist in the mouse event event. Any grabbed view can now give up the grab to another view or maintaining still grabbed by using the events OnMouseGrab/OnMouseUngrab. If a view is doing that for moving a object from one place to another and want to finished the job, other view that can grab the mouse will try to get it, if the current grabbed view want or not. It may have priority over the others.

**Comment by tig** [comment]
To illustrate what I think is wrong with the current Terminal.Gui (and v1) API in this regard, consider this class.

```cs
    public class MouseDemo : View
    {
        public MouseDemo ()
        {
            CanFocus = true;
            MouseEvent += (s, e) =>
                          {
                              if (e.MouseEvent.Flags.HasFlag (MouseFlags.Button1Pressed))
                              {
                                  ColorScheme = Colors.ColorSchemes ["Toplevel"];
                              }
                              if (e.MouseEvent.Flags.HasFlag (MouseFlags.Button1Released))
                              {
                                  ColorScheme = Colors.ColorSchemes ["Dialog"];
                              }
                          };
            MouseLeave += (s, e) => { ColorScheme = Colors.ColorSchemes ["Dialog"]; };
        }
    }
```

I want this view to show the "TopLevel" colorscheme while button1 is pressed; when button1 is released, or the mouse leaves the view, the color should go back to "Dialog".

Here's how it behaves now. Note that if I press button1 outside of the view and drag into the view, the color changes to "Toplevel". This is not what I want. The current API provides me no way of knowing that the `Button1Pressed` flag was set while the mouse was outsdie of my view.

![5DhZHvn 1](https://github.com/gui-cs/Terminal.Gui/assets/585482/8fc7095a-8c99-41af-b7c1-ff5bc6ad0a9b)

Now, I can work around this like this, but this is awful.

```cs
    public class MouseDemo : View
    {
        private bool _button1PressedOnEnter = false;
        public MouseDemo ()
        {
            CanFocus = true;
            MouseEvent += (s, e) =>
                          {
                              if (e.MouseEvent.Flags.HasFlag (MouseFlags.Button1Pressed))
                              {
                                  if (!_button1PressedOnEnter)
                                  {
                                      ColorScheme = Colors.ColorSchemes ["Toplevel"];
                                  }
                              }
                              if (e.MouseEvent.Flags.HasFlag (MouseFlags.Button1Released))
                              {
                                  ColorScheme = Colors.ColorSchemes ["Dialog"];
                                  _button1PressedOnEnter = false;
                              }
                          };
            MouseLeave += (s, e) =>
                          {
                              ColorScheme = Colors.ColorSchemes ["Dialog"];
                              _button1PressedOnEnter = false;
                          };
            MouseEnter += (s, e) =>
                          {
                              _button1PressedOnEnter = e.MouseEvent.Flags.HasFlag (MouseFlags.Button1Pressed);
                          };
        }
    }
```

I think there should be a `View.MouseButton1Pressed` and `View.MouseButton1Released` event.

**Comment by tig** [comment]
FWIW, the impact of this behavior can be seen (in v1 and v2) in the`Windows & FrameViews` scenario.

Press button1 while in the Bounds of `1 - Windows Loop...` and drag up.

![ENpcvt9 1](https://github.com/gui-cs/Terminal.Gui/assets/585482/b5140d28-0cb8-4e7a-8c0e-9736fa485fc8)
